lighthouse-core/report/v2/renderer/report-features.js

```diff
@@ -0,0 +1,326 @@
+/**
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/* globals self URL Blob Logger CustomEvent */
+
+class ReportUIFeatures {
+
+  /**
+   * @param {!DOM} dom
+   */
+  constructor(dom) {
+    this.json = null;
+    /** @private {!DOM} */
+    this._dom = dom;
+    /** @private {!Document} */
+    this._document = this._dom.document();
+    /** @private {boolean} */
+    this._copyAttempt = false;
+    /** @type {Element} **/
+    this.exportButton = null;
+    /** @type {Logger} **/
+    this.logger = null;
+
+    this.onCopy = this.onCopy.bind(this);
+    this.onExportButtonClick = this.onExportButtonClick.bind(this);
+    this.onExport = this.onExport.bind(this);
+    this.onKeyDown = this.onKeyDown.bind(this);
+    this.printShortCutDetect = this.printShortCutDetect.bind(this);
+  }
+
+  /**
+   * Adds export button, print, and other functionality to the report. The method
+   * should be called whenever the report needs to be re-rendered.
+   * @param {!ReportRenderer.ReportJSON} report
+   * @param {!Element} container Parent element to render the report into.
+   */
+  initFeatures(report, container) {
+    this.json = report;
+    this._setupLogger(container);
+    this._setupExportButton();
+    this._setUpCollapseDetailsAfterPrinting();
+    this._resetUIState();
+    this._document.addEventListener('keydown', this.printShortCutDetect);
+    this._document.addEventListener('copy', this.onCopy);
+  }
+
+  /**
+   * Fires a custom DOM event on target.
+   * @param {string} name Name of the event.
+   * @param {!Document|!Element} target DOM node to fire the event on.
+   * @param {Object<{detail: Object<string, *>}>=} detail Custom data to include.
```

Comment:
Not exactly sure where this is going, but from the usage in `onCopyButtonClick`, at least, this could be `@param {{cmd: string, fields: !Object<string, *>}=} detail`, but not sure if that's the intention